### PR TITLE
Bugfix/corbett/face collision

### LINKED ISSUE #2222: incorrect assumption (3-shared nodes) about element-face connectivity for prisms
**Describe the bug**

Incomplete CellElementRegion definition in the input XML file makes GEOSX crash.

As a user I expect GEOSX to return a meaningful error and not crash.

The result of the dynamic_cast done in getGroup function called line 493 of src/coreComponents/mesh/generators/CellBlockManager.cpp should be checked.

**Crashing ranks** 
 
Received signal 11: Segmentation fault

Frame 0: /lib64/libc.so.6
Frame 1: __dynamic_cast
Frame 2: geosx::CellBlock& geosx::dynamicCast<geosx::CellBlock&, geosx::dataRepository::Group>(geosx::dataRepository::Group&)
Frame 3: geosx::CellBlock& geosx::dataRepository::Group::getGroup<geosx::CellBlock, int>(int const&)
Frame 4: geosx::fillElementToEdgesOfCellBlocks(LvArray::ArrayOfArraysView<int const, int const, true, LvArray::ChaiBuffer> const&, geosx::dataRepository::Group&)
Frame 5: geosx::CellBlockManager::buildMaps()
Frame 6: geosx::VTKMeshGenerator::generateMesh(geosx::DomainPartition&)
Frame 7: geosx::MeshManager::generateMeshes(geosx::DomainPartition&)
Frame 8: geosx::ProblemManager::generateMesh()
Frame 9: geosx::ProblemManager::problemSetup()
Frame 10: geosx::GeosxState::initializeDataRepository()
Frame 11: main
Frame 12: __libc_start_main
Frame 13: /workrd/SCR/GEOSX/install/gcc8/GEOSX-a31068c/bin/geosx

**Other ranks**

Failure when gathering information

Frame 6: geosx::vtk::printMeshStatistics(vtkDataSet&, std::map<geosx::ElementType, std::unordered_map<int, std::vector<long long, std::allocator<long long> >, std::hash<int>, std::equal_to<int>, std::allocator<std::pair<int const, std::vector<long long, std::allocator<long long> > > > >, std::less<geosx::ElementType>, std::allocator<std::pair<geosx::ElementType const, std::unordered_map<int, std::vector<long long, std::allocator<long long> >, std::hash<int>, std::equal_to<int>, std::allocator<std::pair<int const, std::vector<long long, std::allocator<long long> > > > > > > > const&, ompi_communicator_t*)
Frame 7: geosx::VTKMeshGenerator::generateMesh(geosx::DomainPartition&)
Frame 8: geosx::MeshManager::generateMeshes(geosx::DomainPartition&)
Frame 9: geosx::ProblemManager::generateMesh()
Frame 10: geosx::ProblemManager::problemSetup()
Frame 11: geosx::GeosxState::initializeDataRepository()
Frame 12: main
Frame 13: __libc_start_main
Frame 14: /workrd/SCR/GEOSX/install/gcc8/GEOSX-a31068c/bin/geosx

### PR DESCRIPTION
**Pull request by corbett5**
fixes #2222

### CODE REVIEW

**Review by klevzoff — APPROVED** [review]
Nice! Out of curiosity, have you measured how this compares perf-wise to current develop for default use case (where 3 nodes are sufficient, e.g. internal mesh)?

File: src/coreComponents/mesh/generators/CellBlockABC.hpp [hunk 1]
```diff
@@ -76,6 +76,12 @@ class CellBlockABC : public dataRepository::Group
    */
   virtual localIndex numFacesPerElement() const = 0;
 
+  /**
+   * @brief Get the maximum number of nodes comprising a face.
+   * @return maximum number of nodes comprising a face.
+   */
+  virtual localIndex maxNodesPerFace() const = 0;
+
```
Comment (TotoGaz) [inline src/coreComponents/mesh/generators/CellBlockABC.hpp:84]: Without diving into the details, I do not think it's required to expose a new function `maxNodesPerFace` into the `CellBlockABC` contract.
This `maxNodesPerFace` information looks like it's a computational information to manage a tricky corner case.
But once the issue's solved, it's not required anymore, is it? (From your PR, it appears that people manipulating `CellBlockABC` abstractions never call `CellBlockABC::maxNodesPerFace`: only the `CellBlock` implementation gets called. Hence `CellBlock::maxNodesPerFace` should not be exposed too broadly.).

I think that if you remove this `virtual` member function, the code will still run fine (expect `CellBlock::maxNodesPerFace` which cannot be defined `override` obviously).

Comment (corbett5) [reply src/coreComponents/mesh/generators/CellBlockABC.hpp:84]: Yes good idea.

File: src/coreComponents/mesh/generators/CellBlockManager.cpp [hunk 2]
```diff
@@ -405,50 +415,90 @@ createLowestNodeToFaces( localIndex const numNodes, const Group & cellBlocks )
         RAJA::atomicInc< parallelHostAtomic >( &counts[ lowestNode ] );
       }
     } );
+
+    totalDuplicateFaces += numFacesPerElement * numElements;
+    totalDuplicateFaceNodes += numElements * cb.maxNodesPerFace();
```
[the remark this answers is not in the document]
Comment (corbett5) [reply src/coreComponents/mesh/generators/CellBlockManager.cpp:420]: Yup, thanks!

**Comment by corbett5** [comment]
Exclusive timings for a `200^3` hex mesh with one rank using an entire Quartz node. The new `buildFaceMaps` takes 50% longer, but IMO that's still pretty good since `calculateShapeFunctionGradients` takes almost 8 times as long. 

This branch
```
geosx::InternalMeshGenerator::generateMesh      1.206378
  geosx::CellBlockManager::buildMaps            0.459340
    geosx::buildEdgeMaps                        0.979828
    geosx::CellBlockManager::buildFaceMaps      1.810109
```

Develop
```
geosx::InternalMeshGenerator::generateMesh      1.228788
  geosx::CellBlockManager::buildMaps            0.415095
    geosx::buildEdgeMaps                        1.042066
    geosx::CellBlockManager::buildFaceMaps      1.239561
```

**Pull request merged**